test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatPushTests.cs

```diff
@@ -0,0 +1,83 @@
+    {
+        [Theory]
+        [InlineData(TestServers.MyGet, nameof(TestServers.MyGet), false)]
+        [InlineData(TestServers.ProGet, nameof(TestServers.ProGet), false)]
+        [InlineData(TestServers.Nexus, nameof(TestServers.Nexus), true)]
+        // [InlineData(TestServers.Artifactory, nameof(TestServers.Artifactory), false)] // Trial license expired!
+        // [InlineData(TestServers.Klondike, nameof(TestServers.Klondike), false)] // 500 Internal Server Error pushing
+        // [InlineData(TestServers.NuGetServer, nameof(TestServers.NuGetServer), false)] // doesn't work (credentials?)
+        public async Task PushToServerSucceeds(string sourceUri, string feedName, bool mustDeleteFirst)
+        {
+            // Arrange
+            using (var packageDir = TestFileSystemUtility.CreateRandomTestFolder())
+            {
+                var packageId = "XPlatPushTests.PushToServerSucceeds";
+                var packageVersion = "1.0.0";
+                var packageFile = await TestPackages.GetRuntimePackageAsync(packageDir, packageId, packageVersion);
+                var configFile = XPlatTestUtils.CopyFuncTestConfig(packageDir);
+                var log = new TestCommandOutputLogger();
+
+                var apiKey = XPlatTestUtils.ReadApiKey(feedName);
+                Assert.False(string.IsNullOrEmpty(apiKey));
+
+                if (mustDeleteFirst)
+                {
+                    DeletePackageBeforePush(packageId, packageVersion, sourceUri, apiKey);
+                }
+
+                var pushArgs = new List<string>()
+                {
+                    "push",
+                    packageFile.FullName,
+                    "--source",
+                    sourceUri,
+                    "--api-key",
+                    apiKey
+                };
+
+                // Act
+                int exitCode = Program.MainInternal(pushArgs.ToArray(), log);
+
+                // Assert
+                Assert.Equal(string.Empty, log.ShowErrors());
+                Assert.Equal(0, exitCode);
+                Assert.Contains($@"PUT {sourceUri}", log.ShowMessages());
+                Assert.Contains("Your package was pushed.", log.ShowMessages());
+            }
+        }
+
+        /// <summary>
+        /// This is called when the package must be deleted before being pushed. It's ok if this
+        /// fails, maybe the package was never pushed.
+        /// </summary>
+        private void DeletePackageBeforePush(string packageId, string packageVersion, string sourceUri, string apiKey)
+        {
+            var log = new TestCommandOutputLogger();
+            var args = new List<string>()
+            {
+                "delete",
+                packageId,
+                packageVersion,
+                "--source",
+                sourceUri,
+                "--api-key",
+                apiKey,
+                "--non-interactive"
+            };
+
+            Program.MainInternal(args.ToArray(), log);
```

Comment:
I added asserts to make sure the DELETE either succeeds or fails with the right output. It can fail if a server gets cleaned up and doesn't have the test package anymore. I don't want that to affect the PUSH because that's what's really being tested.

---
In reply to: [68092607](https://github.com/NuGet/NuGet.Client/pull/706#discussion_r68092607) [](ancestors = 68092607)